// node-opcua/_generated_/_auto_generated_AnonymousIdentityToken.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:08.802Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/AnonymousIdentityToken_schema").AnonymousIdentityToken_Schema;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * A token representing an anonymous user.
 * 
 * @class AnonymousIdentityToken
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.policyId] {String} The policy id specified in a user token policy for the endpoint being used.
 */
function AnonymousIdentityToken(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * The policy id specified in a user token policy for the endpoint being used.
      * @property policyId
      * @type {String}
      */
    self.policyId = initialize_field(schema.fields[0], options.policyId);

   // Object.preventExtensions(self);
}
util.inherits(AnonymousIdentityToken,BaseUAObject);
AnonymousIdentityToken.prototype.encodingDefaultBinary = makeExpandedNodeId(321,0);
AnonymousIdentityToken.prototype.encodingDefaultXml = makeExpandedNodeId(320,0);
AnonymousIdentityToken.prototype._schema = schema;

var encode_String = _defaultTypeMap.String.encode;
var decode_String = _defaultTypeMap.String.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
AnonymousIdentityToken.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_String(this.policyId,stream);
};
AnonymousIdentityToken.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.policyId = decode_String(stream,options);
};
AnonymousIdentityToken.possibleFields = (function() {
    return [
        "policyId"
    ];
})();


exports.AnonymousIdentityToken = AnonymousIdentityToken;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("AnonymousIdentityToken",AnonymousIdentityToken);
