// node-opcua/_generated_/_auto_generated_QueryNextRequest.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:09.282Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/QueryNextRequest_schema").QueryNextRequest_Schema;
var RequestHeader = require("./_auto_generated_RequestHeader").RequestHeader;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class QueryNextRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} A standard header included in all requests sent to a server.
 * @param  [options.releaseContinuationPoint] {Boolean} 
 * @param  [options.continuationPoint] {ContinuationPoint} 
 */
function QueryNextRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * A standard header included in all requests sent to a server.
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * 
      * @property releaseContinuationPoint
      * @type {Boolean}
      */
    self.releaseContinuationPoint = initialize_field(schema.fields[1], options.releaseContinuationPoint);

    /**
      * 
      * @property continuationPoint
      * @type {ContinuationPoint}
      */
    self.continuationPoint = initialize_field(schema.fields[2], options.continuationPoint);

   // Object.preventExtensions(self);
}
util.inherits(QueryNextRequest,BaseUAObject);
QueryNextRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(621,0);
QueryNextRequest.prototype.encodingDefaultXml = makeExpandedNodeId(620,0);
QueryNextRequest.prototype._schema = schema;

var encode_Boolean = _defaultTypeMap.Boolean.encode;
var decode_Boolean = _defaultTypeMap.Boolean.decode;
var encode_ContinuationPoint = _defaultTypeMap.ContinuationPoint.encode;
var decode_ContinuationPoint = _defaultTypeMap.ContinuationPoint.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
QueryNextRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encode_Boolean(this.releaseContinuationPoint,stream);
    encode_ContinuationPoint(this.continuationPoint,stream);
};
QueryNextRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.releaseContinuationPoint = decode_Boolean(stream,options);
    this.continuationPoint = decode_ContinuationPoint(stream,options);
};
QueryNextRequest.possibleFields = (function() {
    return [
        "requestHeader",
         "releaseContinuationPoint",
         "continuationPoint"
    ];
})();


exports.QueryNextRequest = QueryNextRequest;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("QueryNextRequest",QueryNextRequest);
