// node-opcua/_generated_/_auto_generated_EventFilterResult.js
// --------- This code has been automatically generated !!! 2018-04-20T14:13:31.483Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/EventFilterResult_schema").EventFilterResult_Schema;
var DiagnosticInfo = require("./_auto_generated_DiagnosticInfo").DiagnosticInfo;
var ContentFilterResult = require("./_auto_generated_ContentFilterResult").ContentFilterResult;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;
var MonitoringFilter = require("./_auto_generated_MonitoringFilter").MonitoringFilter;

/**
 * 
 * @class EventFilterResult
 * @constructor
 * @extends MonitoringFilter
 * @param  options {Object}
 * @param  [options.selectClauseResults] {StatusCode[]} 
 * @param  [options.selectClauseDiagnosticInfos] {DiagnosticInfo[]} 
 * @param  [options.whereClauseResult] {ContentFilterResult} 
 */
function EventFilterResult(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    MonitoringFilter.call(this,options);

    /**
      * 
      * @property selectClauseResults
      * @type {StatusCode[]}
      */
    self.selectClauseResults = initialize_field_array(schema.fields[0], options.selectClauseResults);

    /**
      * 
      * @property selectClauseDiagnosticInfos
      * @type {DiagnosticInfo[]}
      */
    self.selectClauseDiagnosticInfos = [];
    if (options.selectClauseDiagnosticInfos) {
        assert(_.isArray(options.selectClauseDiagnosticInfos));
        self.selectClauseDiagnosticInfos = options.selectClauseDiagnosticInfos.map(function(e){ return new DiagnosticInfo(e); } );
    }

    /**
      * 
      * @property whereClauseResult
      * @type {ContentFilterResult}
      */
    self.whereClauseResult =  new ContentFilterResult( options.whereClauseResult);

   // Object.preventExtensions(self);
}
util.inherits(EventFilterResult,MonitoringFilter);
EventFilterResult.prototype.encodingDefaultBinary = makeExpandedNodeId(736,0);
EventFilterResult.prototype.encodingDefaultXml = makeExpandedNodeId(735,0);
EventFilterResult.prototype._schema = schema;

var encode_StatusCode = _defaultTypeMap.StatusCode.encode;
var decode_StatusCode = _defaultTypeMap.StatusCode.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
EventFilterResult.prototype.encode = function(stream,options) {
    // call base class implementation first
    MonitoringFilter.prototype.encode.call(this,stream,options);
    encodeArray(this.selectClauseResults, stream, encode_StatusCode);
    encodeArray(this.selectClauseDiagnosticInfos,stream,function(obj,stream){ obj.encode(stream,options); }); 
   this.whereClauseResult.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
EventFilterResult.prototype.decode = function(stream,options) {
    // call base class implementation first
    MonitoringFilter.prototype.decode.call(this,stream,options);
    this.selectClauseResults = decodeArray(stream, decode_StatusCode);
    this.selectClauseDiagnosticInfos = decodeArray(stream, function(stream) { 
       var obj = new DiagnosticInfo();
       obj.decode(stream,options);
       return obj; 
    });
    this.whereClauseResult.decode(stream,options);
};
EventFilterResult.possibleFields = [
  "selectClauseResults",
         "selectClauseDiagnosticInfos",
         "whereClauseResult"
];


exports.EventFilterResult = EventFilterResult;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("EventFilterResult",EventFilterResult);
