// node-opcua/_generated_/_auto_generated_RelativePath.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:09.054Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/RelativePath_schema").RelativePath_Schema;
var RelativePathElement = require("./_auto_generated_RelativePathElement").RelativePathElement;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class RelativePath
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.elements] {RelativePathElement[]} 
 */
function RelativePath(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property elements
      * @type {RelativePathElement[]}
      */
    self.elements = [];
    if (options.elements) {
        assert(_.isArray(options.elements));
        self.elements = options.elements.map(function(e){ return new RelativePathElement(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(RelativePath,BaseUAObject);
RelativePath.prototype.encodingDefaultBinary = makeExpandedNodeId(542,0);
RelativePath.prototype.encodingDefaultXml = makeExpandedNodeId(541,0);
RelativePath.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
RelativePath.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encodeArray(this.elements,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
RelativePath.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.elements = decodeArray(stream, function(stream) { 
       var obj = new RelativePathElement();
       obj.decode(stream,options);
       return obj; 
    });
};
RelativePath.possibleFields = (function() {
    return [
        "elements"
    ];
})();


exports.RelativePath = RelativePath;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("RelativePath",RelativePath);
