// node-opcua/_generated_/_auto_generated_DeleteReferencesRequest.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:09.252Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/DeleteReferencesRequest_schema").DeleteReferencesRequest_Schema;
var RequestHeader = require("./_auto_generated_RequestHeader").RequestHeader;
var DeleteReferencesItem = require("./_auto_generated_DeleteReferencesItem").DeleteReferencesItem;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class DeleteReferencesRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} A standard header included in all requests sent to a server.
 * @param  [options.referencesToDelete] {DeleteReferencesItem[]}  
 */
function DeleteReferencesRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * A standard header included in all requests sent to a server.
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      *  
      * @property referencesToDelete
      * @type {DeleteReferencesItem[]}
      */
    self.referencesToDelete = [];
    if (options.referencesToDelete) {
        assert(_.isArray(options.referencesToDelete));
        self.referencesToDelete = options.referencesToDelete.map(function(e){ return new DeleteReferencesItem(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(DeleteReferencesRequest,BaseUAObject);
DeleteReferencesRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(506,0);
DeleteReferencesRequest.prototype.encodingDefaultXml = makeExpandedNodeId(505,0);
DeleteReferencesRequest.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
DeleteReferencesRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encodeArray(this.referencesToDelete,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
DeleteReferencesRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.referencesToDelete = decodeArray(stream, function(stream) { 
       var obj = new DeleteReferencesItem();
       obj.decode(stream,options);
       return obj; 
    });
};
DeleteReferencesRequest.possibleFields = (function() {
    return [
        "requestHeader",
         "referencesToDelete"
    ];
})();


exports.DeleteReferencesRequest = DeleteReferencesRequest;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("DeleteReferencesRequest",DeleteReferencesRequest);
