// node-opcua/_generated_/_auto_generated_EndpointDescription.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:08.729Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/EndpointDescription_schema").EndpointDescription_Schema;
var ApplicationDescription = require("./_auto_generated_ApplicationDescription").ApplicationDescription;
var UserTokenPolicy = require("./_auto_generated_UserTokenPolicy").UserTokenPolicy;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

//## Define special behavior for Enumeration
var _enum_properties = {
    "securityMode": {
        hidden: false,
        enumerable: true,
        configurable: false,
            get: function() {
                return this.$securityMode;
            },
            set: function(value) {
                var coercedValue = _enumerations.MessageSecurityMode.typedEnum.get(value);
                /* istanbul ignore next */
                if ( coercedValue === undefined || coercedValue === null) {
                      throw new Error("value cannot be coerced to MessageSecurityMode: " + value);
                }
                this.$securityMode = coercedValue;
            }
    },
    "$securityMode": {
         configurable: false,
         value: 'aaaa',
         hidden: true,
         writable: true,
         enumerable: false
    },
};


/**
 * 
 * @class EndpointDescription
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.endpointUrl] {String} 
 * @param  [options.server] {ApplicationDescription} 
 * @param  [options.serverCertificate] {ByteString} 
 * @param  [options.securityMode] {MessageSecurityMode} 
 * @param  [options.securityPolicyUri] {String} 
 * @param  [options.userIdentityTokens] {UserTokenPolicy[]} 
 * @param  [options.transportProfileUri] {String} 
 * @param  [options.securityLevel] {Byte} 
 */
function EndpointDescription(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property endpointUrl
      * @type {String}
      */
    self.endpointUrl = initialize_field(schema.fields[0], options.endpointUrl);

    /**
      * 
      * @property server
      * @type {ApplicationDescription}
      */
    self.server =  new ApplicationDescription( options.server);

    /**
      * 
      * @property serverCertificate
      * @type {ByteString}
      */
    self.serverCertificate = initialize_field(schema.fields[2], options.serverCertificate);

    /**
      * 
      * @property securityMode
      * @type {MessageSecurityMode}
      */
    self.securityMode = initialize_field(schema.fields[3], options.securityMode);

    /**
      * 
      * @property securityPolicyUri
      * @type {String}
      */
    self.securityPolicyUri = initialize_field(schema.fields[4], options.securityPolicyUri);

    /**
      * 
      * @property userIdentityTokens
      * @type {UserTokenPolicy[]}
      */
    self.userIdentityTokens = [];
    if (options.userIdentityTokens) {
        assert(_.isArray(options.userIdentityTokens));
        self.userIdentityTokens = options.userIdentityTokens.map(function(e){ return new UserTokenPolicy(e); } );
    }

    /**
      * 
      * @property transportProfileUri
      * @type {String}
      */
    self.transportProfileUri = initialize_field(schema.fields[6], options.transportProfileUri);

    /**
      * 
      * @property securityLevel
      * @type {Byte}
      */
    self.securityLevel = initialize_field(schema.fields[7], options.securityLevel);

   // Object.preventExtensions(self);
}
util.inherits(EndpointDescription,BaseUAObject);

//define enumeration properties
Object.defineProperties(EndpointDescription.prototype,_enum_properties);
EndpointDescription.prototype.encodingDefaultBinary = makeExpandedNodeId(314,0);
EndpointDescription.prototype.encodingDefaultXml = makeExpandedNodeId(313,0);
EndpointDescription.prototype._schema = schema;

var encode_String = _defaultTypeMap.String.encode;
var decode_String = _defaultTypeMap.String.decode;
var encode_ByteString = _defaultTypeMap.ByteString.encode;
var decode_ByteString = _defaultTypeMap.ByteString.decode;
var encode_MessageSecurityMode = _enumerations.MessageSecurityMode.encode;
var decode_MessageSecurityMode = _enumerations.MessageSecurityMode.decode;
var encode_Byte = _defaultTypeMap.Byte.encode;
var decode_Byte = _defaultTypeMap.Byte.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
EndpointDescription.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_String(this.endpointUrl,stream);
   this.server.encode(stream,options);
    encode_ByteString(this.serverCertificate,stream);
    encode_MessageSecurityMode(this.securityMode,stream);
    encode_String(this.securityPolicyUri,stream);
    encodeArray(this.userIdentityTokens,stream,function(obj,stream){ obj.encode(stream,options); }); 
    encode_String(this.transportProfileUri,stream);
    encode_Byte(this.securityLevel,stream);
};
EndpointDescription.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.endpointUrl = decode_String(stream,options);
    this.server.decode(stream,options);
    this.serverCertificate = decode_ByteString(stream,options);
    this.securityMode = decode_MessageSecurityMode(stream,options);
    this.securityPolicyUri = decode_String(stream,options);
    this.userIdentityTokens = decodeArray(stream, function(stream) { 
       var obj = new UserTokenPolicy();
       obj.decode(stream,options);
       return obj; 
    });
    this.transportProfileUri = decode_String(stream,options);
    this.securityLevel = decode_Byte(stream,options);
};
EndpointDescription.possibleFields = (function() {
    return [
        "endpointUrl",
         "server",
         "serverCertificate",
         "securityMode",
         "securityPolicyUri",
         "userIdentityTokens",
         "transportProfileUri",
         "securityLevel"
    ];
})();


exports.EndpointDescription = EndpointDescription;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("EndpointDescription",EndpointDescription);
