// node-opcua/_generated_/_auto_generated_ModifySubscriptionResponse.js
// --------- This code has been automatically generated !!! 2018-04-20T14:13:31.248Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ModifySubscriptionResponse_schema").ModifySubscriptionResponse_Schema;
var ResponseHeader = require("./_auto_generated_ResponseHeader").ResponseHeader;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ModifySubscriptionResponse
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.responseHeader] {ResponseHeader} 
 * @param  [options.revisedPublishingInterval] {Duration} 
 * @param  [options.revisedLifetimeCount] {Counter} 
 * @param  [options.revisedMaxKeepAliveCount] {Counter} 
 */
function ModifySubscriptionResponse(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property responseHeader
      * @type {ResponseHeader}
      */
    self.responseHeader =  new ResponseHeader( options.responseHeader);

    /**
      * 
      * @property revisedPublishingInterval
      * @type {Duration}
      */
    self.revisedPublishingInterval = initialize_field(schema.fields[1], options.revisedPublishingInterval);

    /**
      * 
      * @property revisedLifetimeCount
      * @type {Counter}
      */
    self.revisedLifetimeCount = initialize_field(schema.fields[2], options.revisedLifetimeCount);

    /**
      * 
      * @property revisedMaxKeepAliveCount
      * @type {Counter}
      */
    self.revisedMaxKeepAliveCount = initialize_field(schema.fields[3], options.revisedMaxKeepAliveCount);

   // Object.preventExtensions(self);
}
util.inherits(ModifySubscriptionResponse,BaseUAObject);
ModifySubscriptionResponse.prototype.encodingDefaultBinary = makeExpandedNodeId(796,0);
ModifySubscriptionResponse.prototype.encodingDefaultXml = makeExpandedNodeId(795,0);
ModifySubscriptionResponse.prototype._schema = schema;

var encode_Duration = _defaultTypeMap.Duration.encode;
var decode_Duration = _defaultTypeMap.Duration.decode;
var encode_Counter = _defaultTypeMap.Counter.encode;
var decode_Counter = _defaultTypeMap.Counter.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ModifySubscriptionResponse.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.responseHeader.encode(stream,options);
    encode_Duration(this.revisedPublishingInterval,stream);
    encode_Counter(this.revisedLifetimeCount,stream);
    encode_Counter(this.revisedMaxKeepAliveCount,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
ModifySubscriptionResponse.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.responseHeader.decode(stream,options);
    this.revisedPublishingInterval = decode_Duration(stream,options);
    this.revisedLifetimeCount = decode_Counter(stream,options);
    this.revisedMaxKeepAliveCount = decode_Counter(stream,options);
};
ModifySubscriptionResponse.possibleFields = [
  "responseHeader",
         "revisedPublishingInterval",
         "revisedLifetimeCount",
         "revisedMaxKeepAliveCount"
];


exports.ModifySubscriptionResponse = ModifySubscriptionResponse;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("ModifySubscriptionResponse",ModifySubscriptionResponse);
