// node-opcua/_generated_/_auto_generated_CloseSessionRequest.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:08.790Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/CloseSessionRequest_schema").CloseSessionRequest_Schema;
var RequestHeader = require("./_auto_generated_RequestHeader").RequestHeader;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * Closes a session with the server.
 * 
 * @class CloseSessionRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} A standard header included in all requests sent to a server.
 * @param  [options.deleteSubscriptions] {Boolean} If TRUE all subscriptions are deleted when the session is closed.
 */
function CloseSessionRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * A standard header included in all requests sent to a server.
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * If TRUE all subscriptions are deleted when the session is closed.
      * @property deleteSubscriptions
      * @type {Boolean}
      */
    self.deleteSubscriptions = initialize_field(schema.fields[1], options.deleteSubscriptions);

   // Object.preventExtensions(self);
}
util.inherits(CloseSessionRequest,BaseUAObject);
CloseSessionRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(473,0);
CloseSessionRequest.prototype.encodingDefaultXml = makeExpandedNodeId(472,0);
CloseSessionRequest.prototype._schema = schema;

var encode_Boolean = _defaultTypeMap.Boolean.encode;
var decode_Boolean = _defaultTypeMap.Boolean.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
CloseSessionRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encode_Boolean(this.deleteSubscriptions,stream);
};
CloseSessionRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.deleteSubscriptions = decode_Boolean(stream,options);
};
CloseSessionRequest.possibleFields = (function() {
    return [
        "requestHeader",
         "deleteSubscriptions"
    ];
})();


exports.CloseSessionRequest = CloseSessionRequest;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("CloseSessionRequest",CloseSessionRequest);
