// node-opcua/_generated_/_auto_generated_FindServersResponse.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:09.193Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/FindServersResponse_schema").FindServersResponse_Schema;
var ResponseHeader = require("./_auto_generated_ResponseHeader").ResponseHeader;
var ApplicationDescription = require("./_auto_generated_ApplicationDescription").ApplicationDescription;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 *  A standard header included in all responses returned by servers.
 * 
 * @class FindServersResponse
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.responseHeader] {ResponseHeader} A standard header included in all responses returned by servers.
 * @param  [options.servers] {ApplicationDescription[]} The servers that met the criteria specified in the request.
 */
function FindServersResponse(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * A standard header included in all responses returned by servers.
      * @property responseHeader
      * @type {ResponseHeader}
      */
    self.responseHeader =  new ResponseHeader( options.responseHeader);

    /**
      * The servers that met the criteria specified in the request.
      * @property servers
      * @type {ApplicationDescription[]}
      */
    self.servers = [];
    if (options.servers) {
        assert(_.isArray(options.servers));
        self.servers = options.servers.map(function(e){ return new ApplicationDescription(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(FindServersResponse,BaseUAObject);
FindServersResponse.prototype.encodingDefaultBinary = makeExpandedNodeId(425,0);
FindServersResponse.prototype.encodingDefaultXml = makeExpandedNodeId(424,0);
FindServersResponse.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
FindServersResponse.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.responseHeader.encode(stream,options);
    encodeArray(this.servers,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
FindServersResponse.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.responseHeader.decode(stream,options);
    this.servers = decodeArray(stream, function(stream) { 
       var obj = new ApplicationDescription();
       obj.decode(stream,options);
       return obj; 
    });
};
FindServersResponse.possibleFields = (function() {
    return [
        "responseHeader",
         "servers"
    ];
})();


exports.FindServersResponse = FindServersResponse;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("FindServersResponse",FindServersResponse);
