// node-opcua/_generated_/_auto_generated_SignedSoftwareCertificate.js
// --------- This code has been automatically generated !!! 2018-04-20T14:13:31.152Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/SignedSoftwareCertificate_schema").SignedSoftwareCertificate_Schema;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class SignedSoftwareCertificate
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.certificateData] {ByteString} The data of the certificate.
 * @param  [options.signature] {ByteString} The digital signature.
 */
function SignedSoftwareCertificate(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * The data of the certificate.
      * @property certificateData
      * @type {ByteString}
      */
    self.certificateData = initialize_field(schema.fields[0], options.certificateData);

    /**
      * The digital signature.
      * @property signature
      * @type {ByteString}
      */
    self.signature = initialize_field(schema.fields[1], options.signature);

   // Object.preventExtensions(self);
}
util.inherits(SignedSoftwareCertificate,BaseUAObject);
SignedSoftwareCertificate.prototype.encodingDefaultBinary = makeExpandedNodeId(346,0);
SignedSoftwareCertificate.prototype.encodingDefaultXml = makeExpandedNodeId(345,0);
SignedSoftwareCertificate.prototype._schema = schema;

var encode_ByteString = _defaultTypeMap.ByteString.encode;
var decode_ByteString = _defaultTypeMap.ByteString.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
SignedSoftwareCertificate.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_ByteString(this.certificateData,stream);
    encode_ByteString(this.signature,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
SignedSoftwareCertificate.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.certificateData = decode_ByteString(stream,options);
    this.signature = decode_ByteString(stream,options);
};
SignedSoftwareCertificate.possibleFields = [
  "certificateData",
         "signature"
];


exports.SignedSoftwareCertificate = SignedSoftwareCertificate;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("SignedSoftwareCertificate",SignedSoftwareCertificate);
