// node-opcua/_generated_/_auto_generated_SequenceHeader.js
// --------- This code has been automatically generated !!! 2018-04-20T14:13:31.353Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/SequenceHeader_schema").SequenceHeader_Schema;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class SequenceHeader
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.sequenceNumber] {UInt32} 
 * @param  [options.requestId] {UInt32} 
 */
function SequenceHeader(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property sequenceNumber
      * @type {UInt32}
      */
    self.sequenceNumber = initialize_field(schema.fields[0], options.sequenceNumber);

    /**
      * 
      * @property requestId
      * @type {UInt32}
      */
    self.requestId = initialize_field(schema.fields[1], options.requestId);

   // Object.preventExtensions(self);
}
util.inherits(SequenceHeader,BaseUAObject);
schema.id = generate_new_id();
SequenceHeader.prototype.encodingDefaultBinary = makeExpandedNodeId(schema.id);
SequenceHeader.prototype._schema = schema;

var encode_UInt32 = _defaultTypeMap.UInt32.encode;
var decode_UInt32 = _defaultTypeMap.UInt32.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
SequenceHeader.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_UInt32(this.sequenceNumber,stream);
    encode_UInt32(this.requestId,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
SequenceHeader.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.sequenceNumber = decode_UInt32(stream,options);
    this.requestId = decode_UInt32(stream,options);
};
SequenceHeader.possibleFields = [
  "sequenceNumber",
         "requestId"
];


exports.SequenceHeader = SequenceHeader;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("SequenceHeader",SequenceHeader);
