// node-opcua/_generated_/_auto_generated_RegisterServerRequest.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:09.178Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/RegisterServerRequest_schema").RegisterServerRequest_Schema;
var RequestHeader = require("./_auto_generated_RequestHeader").RequestHeader;
var RegisteredServer = require("./_auto_generated_RegisteredServer").RegisteredServer;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * Registers a server with the discovery server.
 * 
 * @class RegisterServerRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} A standard header included in all requests sent to a server.
 * @param  [options.server] {RegisteredServer} The server to register.
 */
function RegisterServerRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * A standard header included in all requests sent to a server.
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * The server to register.
      * @property server
      * @type {RegisteredServer}
      */
    self.server =  new RegisteredServer( options.server);

   // Object.preventExtensions(self);
}
util.inherits(RegisterServerRequest,BaseUAObject);
RegisterServerRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(437,0);
RegisterServerRequest.prototype.encodingDefaultXml = makeExpandedNodeId(436,0);
RegisterServerRequest.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
RegisterServerRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
   this.server.encode(stream,options);
};
RegisterServerRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.server.decode(stream,options);
};
RegisterServerRequest.possibleFields = (function() {
    return [
        "requestHeader",
         "server"
    ];
})();


exports.RegisterServerRequest = RegisterServerRequest;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("RegisterServerRequest",RegisterServerRequest);
