// node-opcua/_generated_/_auto_generated_DeleteSubscriptionsRequest.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:08.924Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/DeleteSubscriptionsRequest_schema").DeleteSubscriptionsRequest_Schema;
var RequestHeader = require("./_auto_generated_RequestHeader").RequestHeader;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class DeleteSubscriptionsRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} 
 * @param  [options.subscriptionIds] {IntegerId[]} 
 */
function DeleteSubscriptionsRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * 
      * @property subscriptionIds
      * @type {IntegerId[]}
      */
    self.subscriptionIds = initialize_field_array(schema.fields[1], options.subscriptionIds);

   // Object.preventExtensions(self);
}
util.inherits(DeleteSubscriptionsRequest,BaseUAObject);
DeleteSubscriptionsRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(847,0);
DeleteSubscriptionsRequest.prototype.encodingDefaultXml = makeExpandedNodeId(846,0);
DeleteSubscriptionsRequest.prototype._schema = schema;

var encode_IntegerId = _defaultTypeMap.IntegerId.encode;
var decode_IntegerId = _defaultTypeMap.IntegerId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
DeleteSubscriptionsRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encodeArray(this.subscriptionIds, stream, encode_IntegerId);
};
DeleteSubscriptionsRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.subscriptionIds = decodeArray(stream, decode_IntegerId);
};
DeleteSubscriptionsRequest.possibleFields = (function() {
    return [
        "requestHeader",
         "subscriptionIds"
    ];
})();


exports.DeleteSubscriptionsRequest = DeleteSubscriptionsRequest;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("DeleteSubscriptionsRequest",DeleteSubscriptionsRequest);
