// node-opcua/_generated_/_auto_generated_AddReferencesItem.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:09.223Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/AddReferencesItem_schema").AddReferencesItem_Schema;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

//## Define special behavior for Enumeration
var _enum_properties = {
    "targetNodeClass": {
        hidden: false,
        enumerable: true,
        configurable: false,
            get: function() {
                return this.$targetNodeClass;
            },
            set: function(value) {
                var coercedValue = _enumerations.NodeClass.typedEnum.get(value);
                /* istanbul ignore next */
                if ( coercedValue === undefined || coercedValue === null) {
                      throw new Error("value cannot be coerced to NodeClass: " + value);
                }
                this.$targetNodeClass = coercedValue;
            }
    },
    "$targetNodeClass": {
         configurable: false,
         value: 'aaaa',
         hidden: true,
         writable: true,
         enumerable: false
    },
};


/**
 * 
 * @class AddReferencesItem
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.sourceNodeId] {NodeId} 
 * @param  [options.referenceTypeId] {NodeId} 
 * @param  [options.isForward] {Boolean} 
 * @param  [options.targetServerUri] {String} 
 * @param  [options.targetNodeId] {ExpandedNodeId} 
 * @param  [options.targetNodeClass] {NodeClass} 
 */
function AddReferencesItem(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property sourceNodeId
      * @type {NodeId}
      */
    self.sourceNodeId = initialize_field(schema.fields[0], options.sourceNodeId);

    /**
      * 
      * @property referenceTypeId
      * @type {NodeId}
      */
    self.referenceTypeId = initialize_field(schema.fields[1], options.referenceTypeId);

    /**
      * 
      * @property isForward
      * @type {Boolean}
      */
    self.isForward = initialize_field(schema.fields[2], options.isForward);

    /**
      * 
      * @property targetServerUri
      * @type {String}
      */
    self.targetServerUri = initialize_field(schema.fields[3], options.targetServerUri);

    /**
      * 
      * @property targetNodeId
      * @type {ExpandedNodeId}
      */
    self.targetNodeId = initialize_field(schema.fields[4], options.targetNodeId);

    /**
      * 
      * @property targetNodeClass
      * @type {NodeClass}
      */
    self.targetNodeClass = initialize_field(schema.fields[5], options.targetNodeClass);

   // Object.preventExtensions(self);
}
util.inherits(AddReferencesItem,BaseUAObject);

//define enumeration properties
Object.defineProperties(AddReferencesItem.prototype,_enum_properties);
AddReferencesItem.prototype.encodingDefaultBinary = makeExpandedNodeId(381,0);
AddReferencesItem.prototype.encodingDefaultXml = makeExpandedNodeId(380,0);
AddReferencesItem.prototype._schema = schema;

var encode_NodeId = _defaultTypeMap.NodeId.encode;
var decode_NodeId = _defaultTypeMap.NodeId.decode;
var encode_Boolean = _defaultTypeMap.Boolean.encode;
var decode_Boolean = _defaultTypeMap.Boolean.decode;
var encode_String = _defaultTypeMap.String.encode;
var decode_String = _defaultTypeMap.String.decode;
var encode_ExpandedNodeId = _defaultTypeMap.ExpandedNodeId.encode;
var decode_ExpandedNodeId = _defaultTypeMap.ExpandedNodeId.decode;
var encode_NodeClass = _enumerations.NodeClass.encode;
var decode_NodeClass = _enumerations.NodeClass.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
AddReferencesItem.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_NodeId(this.sourceNodeId,stream);
    encode_NodeId(this.referenceTypeId,stream);
    encode_Boolean(this.isForward,stream);
    encode_String(this.targetServerUri,stream);
    encode_ExpandedNodeId(this.targetNodeId,stream);
    encode_NodeClass(this.targetNodeClass,stream);
};
AddReferencesItem.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.sourceNodeId = decode_NodeId(stream,options);
    this.referenceTypeId = decode_NodeId(stream,options);
    this.isForward = decode_Boolean(stream,options);
    this.targetServerUri = decode_String(stream,options);
    this.targetNodeId = decode_ExpandedNodeId(stream,options);
    this.targetNodeClass = decode_NodeClass(stream,options);
};
AddReferencesItem.possibleFields = (function() {
    return [
        "sourceNodeId",
         "referenceTypeId",
         "isForward",
         "targetServerUri",
         "targetNodeId",
         "targetNodeClass"
    ];
})();


exports.AddReferencesItem = AddReferencesItem;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("AddReferencesItem",AddReferencesItem);
