// node-opcua/_generated_/_auto_generated_ReferenceDescription.js
// --------- This code has been automatically generated !!! 2018-04-20T14:13:31.070Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ReferenceDescription_schema").ReferenceDescription_Schema;
var QualifiedName = require("./_auto_generated_QualifiedName").QualifiedName;
var LocalizedText = require("./_auto_generated_LocalizedText").LocalizedText;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * The description of a reference.
 * 
 * @class ReferenceDescription
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.referenceTypeId] {NodeId} The type of references.
 * @param  [options.isForward] {Boolean} TRUE if the reference is a forward reference.
 * @param  [options.nodeId] {ExpandedNodeId} The id of the target node.
 * @param  [options.browseName] {QualifiedName} The browse name of the target node.
 * @param  [options.displayName] {LocalizedText} The display name of the target node.
 * @param  [options.nodeClass = 0] {NodeClass} The node class of the target node.
 * @param  [options.typeDefinition] {ExpandedNodeId} The type definition of the target node.
 */
function ReferenceDescription(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * The type of references.
      * @property referenceTypeId
      * @type {NodeId}
      */
    self.referenceTypeId = initialize_field(schema.fields[0], options.referenceTypeId);

    /**
      * TRUE if the reference is a forward reference.
      * @property isForward
      * @type {Boolean}
      */
    self.isForward = initialize_field(schema.fields[1], options.isForward);

    /**
      * The id of the target node.
      * @property nodeId
      * @type {ExpandedNodeId}
      */
    self.nodeId = initialize_field(schema.fields[2], options.nodeId);

    /**
      * The browse name of the target node.
      * @property browseName
      * @type {QualifiedName}
      */
    self.browseName =  new QualifiedName( options.browseName);

    /**
      * The display name of the target node.
      * @property displayName
      * @type {LocalizedText}
      */
    self.displayName =  new LocalizedText( options.displayName);

    /**
      * The node class of the target node.
      * @property nodeClass
      * @type {NodeClass}
      * @default  0
      */
    self.setNodeClass(initialize_field(schema.fields[5], options.nodeClass));

    /**
      * The type definition of the target node.
      * @property typeDefinition
      * @type {ExpandedNodeId}
      */
    self.typeDefinition = initialize_field(schema.fields[6], options.typeDefinition);

   // Object.preventExtensions(self);
}
util.inherits(ReferenceDescription,BaseUAObject);

//## Define Enumeration setters
ReferenceDescription.prototype.setNodeClass = function(value) {
   var coercedValue = _enumerations.NodeClass.typedEnum.get(value);
   /* istanbul ignore next */
   if (coercedValue === undefined || coercedValue === null) {
      throw new Error("value cannot be coerced to NodeClass: " + value);
   }
   this.nodeClass = coercedValue;
};
schema.id = generate_new_id();
ReferenceDescription.prototype.encodingDefaultBinary = makeExpandedNodeId(schema.id);
ReferenceDescription.prototype._schema = schema;

var encode_NodeId = _defaultTypeMap.NodeId.encode;
var decode_NodeId = _defaultTypeMap.NodeId.decode;
var encode_Boolean = _defaultTypeMap.Boolean.encode;
var decode_Boolean = _defaultTypeMap.Boolean.decode;
var encode_ExpandedNodeId = _defaultTypeMap.ExpandedNodeId.encode;
var decode_ExpandedNodeId = _defaultTypeMap.ExpandedNodeId.decode;
var encode_NodeClass = _enumerations.NodeClass.encode;
var decode_NodeClass = _enumerations.NodeClass.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ReferenceDescription.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_NodeId(this.referenceTypeId,stream);
    encode_Boolean(this.isForward,stream);
    encode_ExpandedNodeId(this.nodeId,stream);
   this.browseName.encode(stream,options);
   this.displayName.encode(stream,options);
    encode_NodeClass(this.nodeClass,stream);
    encode_ExpandedNodeId(this.typeDefinition,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
ReferenceDescription.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.referenceTypeId = decode_NodeId(stream,options);
    this.isForward = decode_Boolean(stream,options);
    this.nodeId = decode_ExpandedNodeId(stream,options);
    this.browseName.decode(stream,options);
    this.displayName.decode(stream,options);
    this.nodeClass = decode_NodeClass(stream,options);
    this.typeDefinition = decode_ExpandedNodeId(stream,options);
};
ReferenceDescription.possibleFields = [
  "referenceTypeId",
         "isForward",
         "nodeId",
         "browseName",
         "displayName",
         "nodeClass",
         "typeDefinition"
];


exports.ReferenceDescription = ReferenceDescription;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("ReferenceDescription",ReferenceDescription);
