// node-opcua/_generated_/_auto_generated_PublishRequest.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:08.891Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/PublishRequest_schema").PublishRequest_Schema;
var RequestHeader = require("./_auto_generated_RequestHeader").RequestHeader;
var SubscriptionAcknowledgement = require("./_auto_generated_SubscriptionAcknowledgement").SubscriptionAcknowledgement;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class PublishRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} 
 * @param  [options.subscriptionAcknowledgements] {SubscriptionAcknowledgement[]} 
 */
function PublishRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * 
      * @property subscriptionAcknowledgements
      * @type {SubscriptionAcknowledgement[]}
      */
    self.subscriptionAcknowledgements = [];
    if (options.subscriptionAcknowledgements) {
        assert(_.isArray(options.subscriptionAcknowledgements));
        self.subscriptionAcknowledgements = options.subscriptionAcknowledgements.map(function(e){ return new SubscriptionAcknowledgement(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(PublishRequest,BaseUAObject);
PublishRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(826,0);
PublishRequest.prototype.encodingDefaultXml = makeExpandedNodeId(825,0);
PublishRequest.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
PublishRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encodeArray(this.subscriptionAcknowledgements,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
PublishRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.subscriptionAcknowledgements = decodeArray(stream, function(stream) { 
       var obj = new SubscriptionAcknowledgement();
       obj.decode(stream,options);
       return obj; 
    });
};
PublishRequest.possibleFields = (function() {
    return [
        "requestHeader",
         "subscriptionAcknowledgements"
    ];
})();


exports.PublishRequest = PublishRequest;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("PublishRequest",PublishRequest);
