// node-opcua/_generated_/_auto_generated_BrowseResult.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:08.692Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/BrowseResult_schema").BrowseResult_Schema;
var ReferenceDescription = require("./_auto_generated_ReferenceDescription").ReferenceDescription;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * The result of a browse operation.
 * 
 * @class BrowseResult
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.statusCode] {StatusCode} A code indicating any error during the operation.
 * @param  [options.continuationPoint = null] {ContinuationPoint} A value that indicates the operation is incomplete and can be continued by calling BrowseNext.
 * @param  [options.references] {ReferenceDescription[]} A list of references that meet the criteria specified in the request.
 */
function BrowseResult(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * A code indicating any error during the operation.
      * @property statusCode
      * @type {StatusCode}
      */
    self.statusCode = initialize_field(schema.fields[0], options.statusCode);

    /**
      * A value that indicates the operation is incomplete and can be continued by calling BrowseNext.
      * @property continuationPoint
      * @type {ContinuationPoint}
      * @default  null
      */
    self.continuationPoint = initialize_field(schema.fields[1], options.continuationPoint);

    /**
      * A list of references that meet the criteria specified in the request.
      * @property references
      * @type {ReferenceDescription[]}
      */
    self.references = [];
    if (options.references) {
        assert(_.isArray(options.references));
        self.references = options.references.map(function(e){ return new ReferenceDescription(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(BrowseResult,BaseUAObject);
schema.id = generate_new_id();
BrowseResult.prototype.encodingDefaultBinary = makeExpandedNodeId(schema.id);
BrowseResult.prototype._schema = schema;

var encode_StatusCode = _defaultTypeMap.StatusCode.encode;
var decode_StatusCode = _defaultTypeMap.StatusCode.decode;
var encode_ContinuationPoint = _defaultTypeMap.ContinuationPoint.encode;
var decode_ContinuationPoint = _defaultTypeMap.ContinuationPoint.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
BrowseResult.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_StatusCode(this.statusCode,stream);
    encode_ContinuationPoint(this.continuationPoint,stream);
    encodeArray(this.references,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
BrowseResult.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.statusCode = decode_StatusCode(stream,options);
    this.continuationPoint = decode_ContinuationPoint(stream,options);
    this.references = decodeArray(stream, function(stream) { 
       var obj = new ReferenceDescription();
       obj.decode(stream,options);
       return obj; 
    });
};
BrowseResult.possibleFields = (function() {
    return [
        "statusCode",
         "continuationPoint",
         "references"
    ];
})();


exports.BrowseResult = BrowseResult;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("BrowseResult",BrowseResult);
