// node-opcua/_generated_/_auto_generated_MonitoredItemCreateRequest.js
// --------- This code has been automatically generated !!! 2018-04-20T14:13:31.253Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/MonitoredItemCreateRequest_schema").MonitoredItemCreateRequest_Schema;
var ReadValueId = require("./_auto_generated_ReadValueId").ReadValueId;
var MonitoringParameters = require("./_auto_generated_MonitoringParameters").MonitoringParameters;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class MonitoredItemCreateRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.itemToMonitor] {ReadValueId} 
 * @param  [options.monitoringMode] {MonitoringMode} 
 * @param  [options.requestedParameters] {MonitoringParameters} 
 */
function MonitoredItemCreateRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property itemToMonitor
      * @type {ReadValueId}
      */
    self.itemToMonitor =  new ReadValueId( options.itemToMonitor);

    /**
      * 
      * @property monitoringMode
      * @type {MonitoringMode}
      */
    self.setMonitoringMode(initialize_field(schema.fields[1], options.monitoringMode));

    /**
      * 
      * @property requestedParameters
      * @type {MonitoringParameters}
      */
    self.requestedParameters =  new MonitoringParameters( options.requestedParameters);

   // Object.preventExtensions(self);
}
util.inherits(MonitoredItemCreateRequest,BaseUAObject);

//## Define Enumeration setters
MonitoredItemCreateRequest.prototype.setMonitoringMode = function(value) {
   var coercedValue = _enumerations.MonitoringMode.typedEnum.get(value);
   /* istanbul ignore next */
   if (coercedValue === undefined || coercedValue === null) {
      throw new Error("value cannot be coerced to MonitoringMode: " + value);
   }
   this.monitoringMode = coercedValue;
};
MonitoredItemCreateRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(745,0);
MonitoredItemCreateRequest.prototype.encodingDefaultXml = makeExpandedNodeId(744,0);
MonitoredItemCreateRequest.prototype._schema = schema;

var encode_MonitoringMode = _enumerations.MonitoringMode.encode;
var decode_MonitoringMode = _enumerations.MonitoringMode.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
MonitoredItemCreateRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.itemToMonitor.encode(stream,options);
    encode_MonitoringMode(this.monitoringMode,stream);
   this.requestedParameters.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
MonitoredItemCreateRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.itemToMonitor.decode(stream,options);
    this.monitoringMode = decode_MonitoringMode(stream,options);
    this.requestedParameters.decode(stream,options);
};
MonitoredItemCreateRequest.possibleFields = [
  "itemToMonitor",
         "monitoringMode",
         "requestedParameters"
];


exports.MonitoredItemCreateRequest = MonitoredItemCreateRequest;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("MonitoredItemCreateRequest",MonitoredItemCreateRequest);
