// node-opcua/_generated_/_auto_generated_AddNodesResult.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:09.214Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/AddNodesResult_schema").AddNodesResult_Schema;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class AddNodesResult
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.statusCode] {StatusCode}  
 * @param  [options.addedNodeId] {NodeId}  
 */
function AddNodesResult(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      *  
      * @property statusCode
      * @type {StatusCode}
      */
    self.statusCode = initialize_field(schema.fields[0], options.statusCode);

    /**
      *  
      * @property addedNodeId
      * @type {NodeId}
      */
    self.addedNodeId = initialize_field(schema.fields[1], options.addedNodeId);

   // Object.preventExtensions(self);
}
util.inherits(AddNodesResult,BaseUAObject);
AddNodesResult.prototype.encodingDefaultBinary = makeExpandedNodeId(485,0);
AddNodesResult.prototype.encodingDefaultXml = makeExpandedNodeId(484,0);
AddNodesResult.prototype._schema = schema;

var encode_StatusCode = _defaultTypeMap.StatusCode.encode;
var decode_StatusCode = _defaultTypeMap.StatusCode.decode;
var encode_NodeId = _defaultTypeMap.NodeId.encode;
var decode_NodeId = _defaultTypeMap.NodeId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
AddNodesResult.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_StatusCode(this.statusCode,stream);
    encode_NodeId(this.addedNodeId,stream);
};
AddNodesResult.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.statusCode = decode_StatusCode(stream,options);
    this.addedNodeId = decode_NodeId(stream,options);
};
AddNodesResult.possibleFields = (function() {
    return [
        "statusCode",
         "addedNodeId"
    ];
})();


exports.AddNodesResult = AddNodesResult;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("AddNodesResult",AddNodesResult);
