// node-opcua/_generated_/_auto_generated_ModifyMonitoredItemsRequest.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:08.949Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ModifyMonitoredItemsRequest_schema").ModifyMonitoredItemsRequest_Schema;
var RequestHeader = require("./_auto_generated_RequestHeader").RequestHeader;
var MonitoredItemModifyRequest = require("./_auto_generated_MonitoredItemModifyRequest").MonitoredItemModifyRequest;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

//## Define special behavior for Enumeration
var _enum_properties = {
    "timestampsToReturn": {
        hidden: false,
        enumerable: true,
        configurable: false,
            get: function() {
                return this.$timestampsToReturn;
            },
            set: function(value) {
                var coercedValue = _enumerations.TimestampsToReturn.typedEnum.get(value);
                /* istanbul ignore next */
                if ( coercedValue === undefined || coercedValue === null) {
                      throw new Error("value cannot be coerced to TimestampsToReturn: " + value);
                }
                this.$timestampsToReturn = coercedValue;
            }
    },
    "$timestampsToReturn": {
         configurable: false,
         value: 'aaaa',
         hidden: true,
         writable: true,
         enumerable: false
    },
};


/**
 * 
 * @class ModifyMonitoredItemsRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} 
 * @param  [options.subscriptionId] {IntegerId} 
 * @param  [options.timestampsToReturn] {TimestampsToReturn} 
 * @param  [options.itemsToModify] {MonitoredItemModifyRequest[]} 
 */
function ModifyMonitoredItemsRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * 
      * @property subscriptionId
      * @type {IntegerId}
      */
    self.subscriptionId = initialize_field(schema.fields[1], options.subscriptionId);

    /**
      * 
      * @property timestampsToReturn
      * @type {TimestampsToReturn}
      */
    self.timestampsToReturn = initialize_field(schema.fields[2], options.timestampsToReturn);

    /**
      * 
      * @property itemsToModify
      * @type {MonitoredItemModifyRequest[]}
      */
    self.itemsToModify = [];
    if (options.itemsToModify) {
        assert(_.isArray(options.itemsToModify));
        self.itemsToModify = options.itemsToModify.map(function(e){ return new MonitoredItemModifyRequest(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(ModifyMonitoredItemsRequest,BaseUAObject);

//define enumeration properties
Object.defineProperties(ModifyMonitoredItemsRequest.prototype,_enum_properties);
ModifyMonitoredItemsRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(763,0);
ModifyMonitoredItemsRequest.prototype.encodingDefaultXml = makeExpandedNodeId(762,0);
ModifyMonitoredItemsRequest.prototype._schema = schema;

var encode_IntegerId = _defaultTypeMap.IntegerId.encode;
var decode_IntegerId = _defaultTypeMap.IntegerId.decode;
var encode_TimestampsToReturn = _enumerations.TimestampsToReturn.encode;
var decode_TimestampsToReturn = _enumerations.TimestampsToReturn.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ModifyMonitoredItemsRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encode_IntegerId(this.subscriptionId,stream);
    encode_TimestampsToReturn(this.timestampsToReturn,stream);
    encodeArray(this.itemsToModify,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
ModifyMonitoredItemsRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.subscriptionId = decode_IntegerId(stream,options);
    this.timestampsToReturn = decode_TimestampsToReturn(stream,options);
    this.itemsToModify = decodeArray(stream, function(stream) { 
       var obj = new MonitoredItemModifyRequest();
       obj.decode(stream,options);
       return obj; 
    });
};
ModifyMonitoredItemsRequest.possibleFields = (function() {
    return [
        "requestHeader",
         "subscriptionId",
         "timestampsToReturn",
         "itemsToModify"
    ];
})();


exports.ModifyMonitoredItemsRequest = ModifyMonitoredItemsRequest;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("ModifyMonitoredItemsRequest",ModifyMonitoredItemsRequest);
