// node-opcua/_generated_/_auto_generated_ContentFilterElementResult.js
// --------- This code has been automatically generated !!! 2018-04-20T14:13:31.479Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ContentFilterElementResult_schema").ContentFilterElementResult_Schema;
var DiagnosticInfo = require("./_auto_generated_DiagnosticInfo").DiagnosticInfo;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ContentFilterElementResult
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.statusCode] {StatusCode} 
 * @param  [options.operandStatusCodes] {StatusCode} 
 * @param  [options.operandDiagnosticInfos] {DiagnosticInfo} 
 */
function ContentFilterElementResult(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property statusCode
      * @type {StatusCode}
      */
    self.statusCode = initialize_field(schema.fields[0], options.statusCode);

    /**
      * 
      * @property operandStatusCodes
      * @type {StatusCode}
      */
    self.operandStatusCodes = initialize_field(schema.fields[1], options.operandStatusCodes);

    /**
      * 
      * @property operandDiagnosticInfos
      * @type {DiagnosticInfo}
      */
    self.operandDiagnosticInfos =  new DiagnosticInfo( options.operandDiagnosticInfos);

   // Object.preventExtensions(self);
}
util.inherits(ContentFilterElementResult,BaseUAObject);
ContentFilterElementResult.prototype.encodingDefaultBinary = makeExpandedNodeId(606,0);
ContentFilterElementResult.prototype.encodingDefaultXml = makeExpandedNodeId(605,0);
ContentFilterElementResult.prototype._schema = schema;

var encode_StatusCode = _defaultTypeMap.StatusCode.encode;
var decode_StatusCode = _defaultTypeMap.StatusCode.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ContentFilterElementResult.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_StatusCode(this.statusCode,stream);
    encode_StatusCode(this.operandStatusCodes,stream);
   this.operandDiagnosticInfos.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
ContentFilterElementResult.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.statusCode = decode_StatusCode(stream,options);
    this.operandStatusCodes = decode_StatusCode(stream,options);
    this.operandDiagnosticInfos.decode(stream,options);
};
ContentFilterElementResult.possibleFields = [
  "statusCode",
         "operandStatusCodes",
         "operandDiagnosticInfos"
];


exports.ContentFilterElementResult = ContentFilterElementResult;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("ContentFilterElementResult",ContentFilterElementResult);
