// node-opcua/_generated_/_auto_generated_RepublishRequest.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:08.901Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/RepublishRequest_schema").RepublishRequest_Schema;
var RequestHeader = require("./_auto_generated_RequestHeader").RequestHeader;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class RepublishRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} 
 * @param  [options.subscriptionId] {IntegerId} 
 * @param  [options.retransmitSequenceNumber] {Counter} 
 */
function RepublishRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * 
      * @property subscriptionId
      * @type {IntegerId}
      */
    self.subscriptionId = initialize_field(schema.fields[1], options.subscriptionId);

    /**
      * 
      * @property retransmitSequenceNumber
      * @type {Counter}
      */
    self.retransmitSequenceNumber = initialize_field(schema.fields[2], options.retransmitSequenceNumber);

   // Object.preventExtensions(self);
}
util.inherits(RepublishRequest,BaseUAObject);
RepublishRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(832,0);
RepublishRequest.prototype.encodingDefaultXml = makeExpandedNodeId(831,0);
RepublishRequest.prototype._schema = schema;

var encode_IntegerId = _defaultTypeMap.IntegerId.encode;
var decode_IntegerId = _defaultTypeMap.IntegerId.decode;
var encode_Counter = _defaultTypeMap.Counter.encode;
var decode_Counter = _defaultTypeMap.Counter.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
RepublishRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encode_IntegerId(this.subscriptionId,stream);
    encode_Counter(this.retransmitSequenceNumber,stream);
};
RepublishRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.subscriptionId = decode_IntegerId(stream,options);
    this.retransmitSequenceNumber = decode_Counter(stream,options);
};
RepublishRequest.possibleFields = (function() {
    return [
        "requestHeader",
         "subscriptionId",
         "retransmitSequenceNumber"
    ];
})();


exports.RepublishRequest = RepublishRequest;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("RepublishRequest",RepublishRequest);
