// node-opcua/_generated_/_auto_generated_DeleteMonitoredItemsRequest.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:08.909Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/DeleteMonitoredItemsRequest_schema").DeleteMonitoredItemsRequest_Schema;
var RequestHeader = require("./_auto_generated_RequestHeader").RequestHeader;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class DeleteMonitoredItemsRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} 
 * @param  [options.subscriptionId] {IntegerId} 
 * @param  [options.monitoredItemIds] {IntegerId[]} 
 */
function DeleteMonitoredItemsRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * 
      * @property subscriptionId
      * @type {IntegerId}
      */
    self.subscriptionId = initialize_field(schema.fields[1], options.subscriptionId);

    /**
      * 
      * @property monitoredItemIds
      * @type {IntegerId[]}
      */
    self.monitoredItemIds = initialize_field_array(schema.fields[2], options.monitoredItemIds);

   // Object.preventExtensions(self);
}
util.inherits(DeleteMonitoredItemsRequest,BaseUAObject);
DeleteMonitoredItemsRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(781,0);
DeleteMonitoredItemsRequest.prototype.encodingDefaultXml = makeExpandedNodeId(780,0);
DeleteMonitoredItemsRequest.prototype._schema = schema;

var encode_IntegerId = _defaultTypeMap.IntegerId.encode;
var decode_IntegerId = _defaultTypeMap.IntegerId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
DeleteMonitoredItemsRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encode_IntegerId(this.subscriptionId,stream);
    encodeArray(this.monitoredItemIds, stream, encode_IntegerId);
};
DeleteMonitoredItemsRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.subscriptionId = decode_IntegerId(stream,options);
    this.monitoredItemIds = decodeArray(stream, decode_IntegerId);
};
DeleteMonitoredItemsRequest.possibleFields = (function() {
    return [
        "requestHeader",
         "subscriptionId",
         "monitoredItemIds"
    ];
})();


exports.DeleteMonitoredItemsRequest = DeleteMonitoredItemsRequest;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("DeleteMonitoredItemsRequest",DeleteMonitoredItemsRequest);
