// node-opcua/_generated_/_auto_generated_SamplingIntervalDiagnostics.js
// --------- This code has been automatically generated !!! 2018-04-20T14:13:34.209Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/39394884f696ff0bf66bacc9a8032cc074e0158e/SamplingIntervalDiagnostics_schema").SamplingIntervalDiagnostics_Schema;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class SamplingIntervalDiagnostics
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.samplingInterval] {Duration} 
 * @param  [options.monitoredItemCount] {UInt32} 
 * @param  [options.maxMonitoredItemCount] {UInt32} 
 * @param  [options.disabledMonitoredItemCount] {UInt32} 
 */
function SamplingIntervalDiagnostics(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property samplingInterval
      * @type {Duration}
      */
    self.samplingInterval = initialize_field(schema.fields[0], options.samplingInterval);

    /**
      * 
      * @property monitoredItemCount
      * @type {UInt32}
      */
    self.monitoredItemCount = initialize_field(schema.fields[1], options.monitoredItemCount);

    /**
      * 
      * @property maxMonitoredItemCount
      * @type {UInt32}
      */
    self.maxMonitoredItemCount = initialize_field(schema.fields[2], options.maxMonitoredItemCount);

    /**
      * 
      * @property disabledMonitoredItemCount
      * @type {UInt32}
      */
    self.disabledMonitoredItemCount = initialize_field(schema.fields[3], options.disabledMonitoredItemCount);

   // Object.preventExtensions(self);
}
util.inherits(SamplingIntervalDiagnostics,BaseUAObject);
SamplingIntervalDiagnostics.prototype.encodingDefaultBinary = makeExpandedNodeId(858,0);
SamplingIntervalDiagnostics.prototype._schema = schema;

var encode_Duration = _defaultTypeMap.Duration.encode;
var decode_Duration = _defaultTypeMap.Duration.decode;
var encode_UInt32 = _defaultTypeMap.UInt32.encode;
var decode_UInt32 = _defaultTypeMap.UInt32.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
SamplingIntervalDiagnostics.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_Duration(this.samplingInterval,stream);
    encode_UInt32(this.monitoredItemCount,stream);
    encode_UInt32(this.maxMonitoredItemCount,stream);
    encode_UInt32(this.disabledMonitoredItemCount,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
SamplingIntervalDiagnostics.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.samplingInterval = decode_Duration(stream,options);
    this.monitoredItemCount = decode_UInt32(stream,options);
    this.maxMonitoredItemCount = decode_UInt32(stream,options);
    this.disabledMonitoredItemCount = decode_UInt32(stream,options);
};
SamplingIntervalDiagnostics.possibleFields = [
  "samplingInterval",
         "monitoredItemCount",
         "maxMonitoredItemCount",
         "disabledMonitoredItemCount"
];


exports.SamplingIntervalDiagnostics = SamplingIntervalDiagnostics;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("SamplingIntervalDiagnostics",SamplingIntervalDiagnostics);
