// node-opcua/_generated_/_auto_generated_WriteRequest.js
// --------- This code has been automatically generated !!! 2018-04-20T14:13:31.436Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/WriteRequest_schema").WriteRequest_Schema;
var RequestHeader = require("./_auto_generated_RequestHeader").RequestHeader;
var WriteValue = require("./_auto_generated_WriteValue").WriteValue;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class WriteRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} 
 * @param  [options.nodesToWrite] {WriteValue[]} 
 */
function WriteRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * 
      * @property nodesToWrite
      * @type {WriteValue[]}
      */
    self.nodesToWrite = [];
    if (options.nodesToWrite) {
        assert(_.isArray(options.nodesToWrite));
        self.nodesToWrite = options.nodesToWrite.map(function(e){ return new WriteValue(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(WriteRequest,BaseUAObject);
WriteRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(673,0);
WriteRequest.prototype.encodingDefaultXml = makeExpandedNodeId(672,0);
WriteRequest.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
WriteRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encodeArray(this.nodesToWrite,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
WriteRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.nodesToWrite = decodeArray(stream, function(stream) { 
       var obj = new WriteValue();
       obj.decode(stream,options);
       return obj; 
    });
};
WriteRequest.possibleFields = [
  "requestHeader",
         "nodesToWrite"
];


exports.WriteRequest = WriteRequest;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("WriteRequest",WriteRequest);
