// node-opcua/_generated_/_auto_generated_ContentFilterResult.js
// --------- This code has been automatically generated !!! 2018-04-20T14:13:31.481Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ContentFilterResult_schema").ContentFilterResult_Schema;
var ContentFilterElementResult = require("./_auto_generated_ContentFilterElementResult").ContentFilterElementResult;
var DiagnosticInfo = require("./_auto_generated_DiagnosticInfo").DiagnosticInfo;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ContentFilterResult
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.elementResults] {ContentFilterElementResult[]} 
 * @param  [options.elementDiagnosticInfos] {DiagnosticInfo[]} 
 */
function ContentFilterResult(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property elementResults
      * @type {ContentFilterElementResult[]}
      */
    self.elementResults = [];
    if (options.elementResults) {
        assert(_.isArray(options.elementResults));
        self.elementResults = options.elementResults.map(function(e){ return new ContentFilterElementResult(e); } );
    }

    /**
      * 
      * @property elementDiagnosticInfos
      * @type {DiagnosticInfo[]}
      */
    self.elementDiagnosticInfos = [];
    if (options.elementDiagnosticInfos) {
        assert(_.isArray(options.elementDiagnosticInfos));
        self.elementDiagnosticInfos = options.elementDiagnosticInfos.map(function(e){ return new DiagnosticInfo(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(ContentFilterResult,BaseUAObject);
ContentFilterResult.prototype.encodingDefaultBinary = makeExpandedNodeId(609,0);
ContentFilterResult.prototype.encodingDefaultXml = makeExpandedNodeId(608,0);
ContentFilterResult.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ContentFilterResult.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encodeArray(this.elementResults,stream,function(obj,stream){ obj.encode(stream,options); }); 
    encodeArray(this.elementDiagnosticInfos,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
ContentFilterResult.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.elementResults = decodeArray(stream, function(stream) { 
       var obj = new ContentFilterElementResult();
       obj.decode(stream,options);
       return obj; 
    });
    this.elementDiagnosticInfos = decodeArray(stream, function(stream) { 
       var obj = new DiagnosticInfo();
       obj.decode(stream,options);
       return obj; 
    });
};
ContentFilterResult.possibleFields = [
  "elementResults",
         "elementDiagnosticInfos"
];


exports.ContentFilterResult = ContentFilterResult;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("ContentFilterResult",ContentFilterResult);
