// node-opcua/_generated_/_auto_generated_ContentFilterElement.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:09.116Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ContentFilterElement_schema").ContentFilterElement_Schema;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

//## Define special behavior for Enumeration
var _enum_properties = {
    "filterOperator": {
        hidden: false,
        enumerable: true,
        configurable: false,
            get: function() {
                return this.$filterOperator;
            },
            set: function(value) {
                var coercedValue = _enumerations.FilterOperator.typedEnum.get(value);
                /* istanbul ignore next */
                if ( coercedValue === undefined || coercedValue === null) {
                      throw new Error("value cannot be coerced to FilterOperator: " + value);
                }
                this.$filterOperator = coercedValue;
            }
    },
    "$filterOperator": {
         configurable: false,
         value: 'aaaa',
         hidden: true,
         writable: true,
         enumerable: false
    },
};


/**
 * 
 * @class ContentFilterElement
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.filterOperator] {FilterOperator} 
 * @param  [options.filterOperands] {ExtensionObject[]} 
 */
function ContentFilterElement(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property filterOperator
      * @type {FilterOperator}
      */
    self.filterOperator = initialize_field(schema.fields[0], options.filterOperator);

    /**
      * 
      * @property filterOperands
      * @type {ExtensionObject[]}
      */
    self.filterOperands = initialize_field_array(schema.fields[1], options.filterOperands);

   // Object.preventExtensions(self);
}
util.inherits(ContentFilterElement,BaseUAObject);

//define enumeration properties
Object.defineProperties(ContentFilterElement.prototype,_enum_properties);
ContentFilterElement.prototype.encodingDefaultBinary = makeExpandedNodeId(585,0);
ContentFilterElement.prototype.encodingDefaultXml = makeExpandedNodeId(584,0);
ContentFilterElement.prototype._schema = schema;

var encode_FilterOperator = _enumerations.FilterOperator.encode;
var decode_FilterOperator = _enumerations.FilterOperator.decode;
var encode_ExtensionObject = _defaultTypeMap.ExtensionObject.encode;
var decode_ExtensionObject = _defaultTypeMap.ExtensionObject.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ContentFilterElement.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_FilterOperator(this.filterOperator,stream);
    encodeArray(this.filterOperands, stream, encode_ExtensionObject);
};
ContentFilterElement.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.filterOperator = decode_FilterOperator(stream,options);
    this.filterOperands = decodeArray(stream, decode_ExtensionObject);
};
ContentFilterElement.possibleFields = (function() {
    return [
        "filterOperator",
         "filterOperands"
    ];
})();


exports.ContentFilterElement = ContentFilterElement;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("ContentFilterElement",ContentFilterElement);
