// node-opcua/_generated_/_auto_generated_CallResponse.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:09.168Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/CallResponse_schema").CallResponse_Schema;
var ResponseHeader = require("./_auto_generated_ResponseHeader").ResponseHeader;
var CallMethodResult = require("./_auto_generated_CallMethodResult").CallMethodResult;
var DiagnosticInfo = require("./_auto_generated_DiagnosticInfo").DiagnosticInfo;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * This Service is used to call (invoke) a list of Methods.
 * 
 * @class CallResponse
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.responseHeader] {ResponseHeader} A standard header included in all responses returned by servers.
 * @param  [options.results] {CallMethodResult[]} Result for the Method calls.
 * @param  [options.diagnosticInfos] {DiagnosticInfo[]} The List of diagnostic information for the statusCode of the results.
 */
function CallResponse(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * A standard header included in all responses returned by servers.
      * @property responseHeader
      * @type {ResponseHeader}
      */
    self.responseHeader =  new ResponseHeader( options.responseHeader);

    /**
      * Result for the Method calls.
      * @property results
      * @type {CallMethodResult[]}
      */
    self.results = [];
    if (options.results) {
        assert(_.isArray(options.results));
        self.results = options.results.map(function(e){ return new CallMethodResult(e); } );
    }

    /**
      * The List of diagnostic information for the statusCode of the results.
      * @property diagnosticInfos
      * @type {DiagnosticInfo[]}
      */
    self.diagnosticInfos = [];
    if (options.diagnosticInfos) {
        assert(_.isArray(options.diagnosticInfos));
        self.diagnosticInfos = options.diagnosticInfos.map(function(e){ return new DiagnosticInfo(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(CallResponse,BaseUAObject);
CallResponse.prototype.encodingDefaultBinary = makeExpandedNodeId(715,0);
CallResponse.prototype.encodingDefaultXml = makeExpandedNodeId(714,0);
CallResponse.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
CallResponse.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.responseHeader.encode(stream,options);
    encodeArray(this.results,stream,function(obj,stream){ obj.encode(stream,options); }); 
    encodeArray(this.diagnosticInfos,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
CallResponse.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.responseHeader.decode(stream,options);
    this.results = decodeArray(stream, function(stream) { 
       var obj = new CallMethodResult();
       obj.decode(stream,options);
       return obj; 
    });
    this.diagnosticInfos = decodeArray(stream, function(stream) { 
       var obj = new DiagnosticInfo();
       obj.decode(stream,options);
       return obj; 
    });
};
CallResponse.possibleFields = (function() {
    return [
        "responseHeader",
         "results",
         "diagnosticInfos"
    ];
})();


exports.CallResponse = CallResponse;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("CallResponse",CallResponse);
