// node-opcua/_generated_/_auto_generated_QualifiedName.js
// --------- This code has been automatically generated !!! 2018-04-20T14:13:30.998Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/QualifiedName_schema").QualifiedName_Schema;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class QualifiedName
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.namespaceIndex] {UInt16} The namespace index
 * @param  [options.name = null] {String} The name
 */
function QualifiedName(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * The namespace index
      * @property namespaceIndex
      * @type {UInt16}
      */
    self.namespaceIndex = initialize_field(schema.fields[0], options.namespaceIndex);

    /**
      * The name
      * @property name
      * @type {String}
      * @default  function () {
            return null;
        }
      */
    self.name = initialize_field(schema.fields[1], options.name);

   // Object.preventExtensions(self);
}
util.inherits(QualifiedName,BaseUAObject);
schema.id = generate_new_id();
QualifiedName.prototype.encodingDefaultBinary = makeExpandedNodeId(schema.id);
QualifiedName.prototype._schema = schema;

var encode_UInt16 = _defaultTypeMap.UInt16.encode;
var decode_UInt16 = _defaultTypeMap.UInt16.decode;
var encode_String = _defaultTypeMap.String.encode;
var decode_String = _defaultTypeMap.String.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
QualifiedName.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_UInt16(this.namespaceIndex,stream);
    encode_String(this.name,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
QualifiedName.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.namespaceIndex = decode_UInt16(stream,options);
    this.name = decode_String(stream,options);
};
QualifiedName.possibleFields = [
  "namespaceIndex",
         "name"
];


exports.QualifiedName = QualifiedName;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("QualifiedName",QualifiedName);
