// node-opcua/_generated_/_auto_generated_EnumValueType.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:08.852Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/EnumValueType_schema").EnumValueType_Schema;
var LocalizedText = require("./_auto_generated_LocalizedText").LocalizedText;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class EnumValueType
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.value] {Int64} The Integer representation of an Enumeration.
 * @param  [options.displayName] {LocalizedText} A human-readable representation of the Value of the Enumeration.
 * @param  [options.description] {LocalizedText} A localized description of the enumeration value. This field can contain an empty string if no description is available
 */
function EnumValueType(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * The Integer representation of an Enumeration.
      * @property value
      * @type {Int64}
      */
    self.value = initialize_field(schema.fields[0], options.value);

    /**
      * A human-readable representation of the Value of the Enumeration.
      * @property displayName
      * @type {LocalizedText}
      */
    self.displayName =  new LocalizedText( options.displayName);

    /**
      * A localized description of the enumeration value. This field can contain an empty string if no description is available
      * @property description
      * @type {LocalizedText}
      */
    self.description =  new LocalizedText( options.description);

   // Object.preventExtensions(self);
}
util.inherits(EnumValueType,BaseUAObject);
EnumValueType.prototype.encodingDefaultBinary = makeExpandedNodeId(8251,0);
EnumValueType.prototype.encodingDefaultXml = makeExpandedNodeId(7616,0);
EnumValueType.prototype._schema = schema;

var encode_Int64 = _defaultTypeMap.Int64.encode;
var decode_Int64 = _defaultTypeMap.Int64.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
EnumValueType.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_Int64(this.value,stream);
   this.displayName.encode(stream,options);
   this.description.encode(stream,options);
};
EnumValueType.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.value = decode_Int64(stream,options);
    this.displayName.decode(stream,options);
    this.description.decode(stream,options);
};
EnumValueType.possibleFields = (function() {
    return [
        "value",
         "displayName",
         "description"
    ];
})();


exports.EnumValueType = EnumValueType;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("EnumValueType",EnumValueType);
