// node-opcua/_generated_/_auto_generated_ModifySubscriptionRequest.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:08.862Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ModifySubscriptionRequest_schema").ModifySubscriptionRequest_Schema;
var RequestHeader = require("./_auto_generated_RequestHeader").RequestHeader;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ModifySubscriptionRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} 
 * @param  [options.subscriptionId] {IntegerId} 
 * @param  [options.requestedPublishingInterval] {Duration} 
 * @param  [options.requestedLifetimeCount] {Counter} 
 * @param  [options.requestedMaxKeepAliveCount] {Counter} 
 * @param  [options.maxNotificationsPerPublish] {Counter} 
 * @param  [options.priority] {Byte} 
 */
function ModifySubscriptionRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * 
      * @property subscriptionId
      * @type {IntegerId}
      */
    self.subscriptionId = initialize_field(schema.fields[1], options.subscriptionId);

    /**
      * 
      * @property requestedPublishingInterval
      * @type {Duration}
      */
    self.requestedPublishingInterval = initialize_field(schema.fields[2], options.requestedPublishingInterval);

    /**
      * 
      * @property requestedLifetimeCount
      * @type {Counter}
      */
    self.requestedLifetimeCount = initialize_field(schema.fields[3], options.requestedLifetimeCount);

    /**
      * 
      * @property requestedMaxKeepAliveCount
      * @type {Counter}
      */
    self.requestedMaxKeepAliveCount = initialize_field(schema.fields[4], options.requestedMaxKeepAliveCount);

    /**
      * 
      * @property maxNotificationsPerPublish
      * @type {Counter}
      */
    self.maxNotificationsPerPublish = initialize_field(schema.fields[5], options.maxNotificationsPerPublish);

    /**
      * 
      * @property priority
      * @type {Byte}
      */
    self.priority = initialize_field(schema.fields[6], options.priority);

   // Object.preventExtensions(self);
}
util.inherits(ModifySubscriptionRequest,BaseUAObject);
ModifySubscriptionRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(793,0);
ModifySubscriptionRequest.prototype.encodingDefaultXml = makeExpandedNodeId(792,0);
ModifySubscriptionRequest.prototype._schema = schema;

var encode_IntegerId = _defaultTypeMap.IntegerId.encode;
var decode_IntegerId = _defaultTypeMap.IntegerId.decode;
var encode_Duration = _defaultTypeMap.Duration.encode;
var decode_Duration = _defaultTypeMap.Duration.decode;
var encode_Counter = _defaultTypeMap.Counter.encode;
var decode_Counter = _defaultTypeMap.Counter.decode;
var encode_Byte = _defaultTypeMap.Byte.encode;
var decode_Byte = _defaultTypeMap.Byte.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ModifySubscriptionRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encode_IntegerId(this.subscriptionId,stream);
    encode_Duration(this.requestedPublishingInterval,stream);
    encode_Counter(this.requestedLifetimeCount,stream);
    encode_Counter(this.requestedMaxKeepAliveCount,stream);
    encode_Counter(this.maxNotificationsPerPublish,stream);
    encode_Byte(this.priority,stream);
};
ModifySubscriptionRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.subscriptionId = decode_IntegerId(stream,options);
    this.requestedPublishingInterval = decode_Duration(stream,options);
    this.requestedLifetimeCount = decode_Counter(stream,options);
    this.requestedMaxKeepAliveCount = decode_Counter(stream,options);
    this.maxNotificationsPerPublish = decode_Counter(stream,options);
    this.priority = decode_Byte(stream,options);
};
ModifySubscriptionRequest.possibleFields = (function() {
    return [
        "requestHeader",
         "subscriptionId",
         "requestedPublishingInterval",
         "requestedLifetimeCount",
         "requestedMaxKeepAliveCount",
         "maxNotificationsPerPublish",
         "priority"
    ];
})();


exports.ModifySubscriptionRequest = ModifySubscriptionRequest;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("ModifySubscriptionRequest",ModifySubscriptionRequest);
