// node-opcua/_generated_/_auto_generated_CallMethodRequest.js
// --------- This code has been automatically generated !!! 2018-04-20T14:13:31.549Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/CallMethodRequest_schema").CallMethodRequest_Schema;
var Variant = require("./_auto_generated_Variant").Variant;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class CallMethodRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.objectId] {NodeId} The NodeId shall be that of the Object or ObjectType that is the source of a HasComponent Reference (or subtype of HasComponent Reference) to the Method specified in methodId.
 * @param  [options.methodId] {NodeId} The NodeId of the Method to invoke.
 * @param  [options.inputArguments] {Variant[]} The list of input argument values.
 */
function CallMethodRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * The NodeId shall be that of the Object or ObjectType that is the source of a HasComponent Reference (or subtype of HasComponent Reference) to the Method specified in methodId.
      * @property objectId
      * @type {NodeId}
      */
    self.objectId = initialize_field(schema.fields[0], options.objectId);

    /**
      * The NodeId of the Method to invoke.
      * @property methodId
      * @type {NodeId}
      */
    self.methodId = initialize_field(schema.fields[1], options.methodId);

    /**
      * The list of input argument values.
      * @property inputArguments
      * @type {Variant[]}
      */
    self.inputArguments = [];
    if (options.inputArguments) {
        assert(_.isArray(options.inputArguments));
        self.inputArguments = options.inputArguments.map(function(e){ return new Variant(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(CallMethodRequest,BaseUAObject);
CallMethodRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(706,0);
CallMethodRequest.prototype.encodingDefaultXml = makeExpandedNodeId(705,0);
CallMethodRequest.prototype._schema = schema;

var encode_NodeId = _defaultTypeMap.NodeId.encode;
var decode_NodeId = _defaultTypeMap.NodeId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
CallMethodRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_NodeId(this.objectId,stream);
    encode_NodeId(this.methodId,stream);
    encodeArray(this.inputArguments,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
CallMethodRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.objectId = decode_NodeId(stream,options);
    this.methodId = decode_NodeId(stream,options);
    this.inputArguments = decodeArray(stream, function(stream) { 
       var obj = new Variant();
       obj.decode(stream,options);
       return obj; 
    });
};
CallMethodRequest.possibleFields = [
  "objectId",
         "methodId",
         "inputArguments"
];


exports.CallMethodRequest = CallMethodRequest;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("CallMethodRequest",CallMethodRequest);
