// node-opcua/_generated_/_auto_generated_AddReferencesRequest.js
// --------- This code has been automatically generated !!! 2018-04-20T14:13:31.588Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/AddReferencesRequest_schema").AddReferencesRequest_Schema;
var RequestHeader = require("./_auto_generated_RequestHeader").RequestHeader;
var AddReferencesItem = require("./_auto_generated_AddReferencesItem").AddReferencesItem;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class AddReferencesRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} A standard header included in all requests sent to a server.
 * @param  [options.referencesToAdd] {AddReferencesItem[]}  
 */
function AddReferencesRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * A standard header included in all requests sent to a server.
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      *  
      * @property referencesToAdd
      * @type {AddReferencesItem[]}
      */
    self.referencesToAdd = [];
    if (options.referencesToAdd) {
        assert(_.isArray(options.referencesToAdd));
        self.referencesToAdd = options.referencesToAdd.map(function(e){ return new AddReferencesItem(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(AddReferencesRequest,BaseUAObject);
AddReferencesRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(494,0);
AddReferencesRequest.prototype.encodingDefaultXml = makeExpandedNodeId(493,0);
AddReferencesRequest.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
AddReferencesRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encodeArray(this.referencesToAdd,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
AddReferencesRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.referencesToAdd = decodeArray(stream, function(stream) { 
       var obj = new AddReferencesItem();
       obj.decode(stream,options);
       return obj; 
    });
};
AddReferencesRequest.possibleFields = [
  "requestHeader",
         "referencesToAdd"
];


exports.AddReferencesRequest = AddReferencesRequest;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("AddReferencesRequest",AddReferencesRequest);
