// node-opcua/_generated_/_auto_generated_MonitoredItemModifyRequest.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:08.943Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/MonitoredItemModifyRequest_schema").MonitoredItemModifyRequest_Schema;
var MonitoringParameters = require("./_auto_generated_MonitoringParameters").MonitoringParameters;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class MonitoredItemModifyRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.monitoredItemId] {IntegerId} 
 * @param  [options.requestedParameters] {MonitoringParameters} 
 */
function MonitoredItemModifyRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property monitoredItemId
      * @type {IntegerId}
      */
    self.monitoredItemId = initialize_field(schema.fields[0], options.monitoredItemId);

    /**
      * 
      * @property requestedParameters
      * @type {MonitoringParameters}
      */
    self.requestedParameters =  new MonitoringParameters( options.requestedParameters);

   // Object.preventExtensions(self);
}
util.inherits(MonitoredItemModifyRequest,BaseUAObject);
MonitoredItemModifyRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(757,0);
MonitoredItemModifyRequest.prototype.encodingDefaultXml = makeExpandedNodeId(756,0);
MonitoredItemModifyRequest.prototype._schema = schema;

var encode_IntegerId = _defaultTypeMap.IntegerId.encode;
var decode_IntegerId = _defaultTypeMap.IntegerId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
MonitoredItemModifyRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_IntegerId(this.monitoredItemId,stream);
   this.requestedParameters.encode(stream,options);
};
MonitoredItemModifyRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.monitoredItemId = decode_IntegerId(stream,options);
    this.requestedParameters.decode(stream,options);
};
MonitoredItemModifyRequest.possibleFields = (function() {
    return [
        "monitoredItemId",
         "requestedParameters"
    ];
})();


exports.MonitoredItemModifyRequest = MonitoredItemModifyRequest;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("MonitoredItemModifyRequest",MonitoredItemModifyRequest);
