// node-opcua/_generated_/_auto_generated_ResponseHeader.js
// --------- This code has been automatically generated !!! 2018-02-01T15:08:08.648Z
"use strict";
require("requirish")._(module);
/**
 * @module opcua.address_space.types
 */
var assert = require("better-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("lib/datamodel/nodeid").makeNodeId;
var schema_helpers =  require("lib/misc/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("lib/misc/factories_builtin_types")._defaultTypeMap;
var ec = require("lib/misc/encode_decode");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = ec.makeExpandedNodeId;
var generate_new_id = require("lib/misc/factories").generate_new_id;
var _enumerations = require("lib/misc/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ResponseHeader_schema").ResponseHeader_Schema;
var DiagnosticInfo = require("./_auto_generated_DiagnosticInfo").DiagnosticInfo;
var BaseUAObject = require("lib/misc/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ResponseHeader
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.timeStamp = Thu Feb 01 2018 16:08:08 GMT+0100 (CET)] {UtcTime} 
 * @param  [options.requestHandle] {IntegerId} 
 * @param  [options.serviceResult] {StatusCode} 
 * @param  [options.serviceDiagnostics] {DiagnosticInfo} The diagnostics associated with the ServiceResult.
 * @param  [options.stringTable] {String[]} 
 * @param  [options.additionalHeader] {ExtensionObject} 
 */
function ResponseHeader(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);

    /**
      * 
      * @property timeStamp
      * @type {UtcTime}
      * @default  function () {
            return new Date();
        }
      */
    self.timeStamp = initialize_field(schema.fields[0], options.timeStamp);

    /**
      * 
      * @property requestHandle
      * @type {IntegerId}
      */
    self.requestHandle = initialize_field(schema.fields[1], options.requestHandle);

    /**
      * 
      * @property serviceResult
      * @type {StatusCode}
      */
    self.serviceResult = initialize_field(schema.fields[2], options.serviceResult);

    /**
      * The diagnostics associated with the ServiceResult.
      * @property serviceDiagnostics
      * @type {DiagnosticInfo}
      */
    self.serviceDiagnostics =  new DiagnosticInfo( options.serviceDiagnostics);

    /**
      * 
      * @property stringTable
      * @type {String[]}
      */
    self.stringTable = initialize_field_array(schema.fields[4], options.stringTable);

    /**
      * 
      * @property additionalHeader
      * @type {ExtensionObject}
      */
    self.additionalHeader = initialize_field(schema.fields[5], options.additionalHeader);

   // Object.preventExtensions(self);
}
util.inherits(ResponseHeader,BaseUAObject);
ResponseHeader.prototype.encodingDefaultBinary = makeExpandedNodeId(394,0);
ResponseHeader.prototype.encodingDefaultXml = makeExpandedNodeId(393,0);
ResponseHeader.prototype._schema = schema;

var encode_UtcTime = _defaultTypeMap.UtcTime.encode;
var decode_UtcTime = _defaultTypeMap.UtcTime.decode;
var encode_IntegerId = _defaultTypeMap.IntegerId.encode;
var decode_IntegerId = _defaultTypeMap.IntegerId.decode;
var encode_StatusCode = _defaultTypeMap.StatusCode.encode;
var decode_StatusCode = _defaultTypeMap.StatusCode.decode;
var encode_String = _defaultTypeMap.String.encode;
var decode_String = _defaultTypeMap.String.decode;
var encode_ExtensionObject = _defaultTypeMap.ExtensionObject.encode;
var decode_ExtensionObject = _defaultTypeMap.ExtensionObject.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ResponseHeader.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_UtcTime(this.timeStamp,stream);
    encode_IntegerId(this.requestHandle,stream);
    encode_StatusCode(this.serviceResult,stream);
   this.serviceDiagnostics.encode(stream,options);
    encodeArray(this.stringTable, stream, encode_String);
    encode_ExtensionObject(this.additionalHeader,stream);
};
ResponseHeader.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.timeStamp = decode_UtcTime(stream,options);
    this.requestHandle = decode_IntegerId(stream,options);
    this.serviceResult = decode_StatusCode(stream,options);
    this.serviceDiagnostics.decode(stream,options);
    this.stringTable = decodeArray(stream, decode_String);
    this.additionalHeader = decode_ExtensionObject(stream,options);
};
ResponseHeader.possibleFields = (function() {
    return [
        "timeStamp",
         "requestHandle",
         "serviceResult",
         "serviceDiagnostics",
         "stringTable",
         "additionalHeader"
    ];
})();


exports.ResponseHeader = ResponseHeader;
var register_class_definition = require("lib/misc/factories_factories").register_class_definition;
register_class_definition("ResponseHeader",ResponseHeader);
